src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs

```diff
@@ -828,12 +828,54 @@ private static void ReadDiagnostic(JsonReader reader, RazorDiagnosticCollection
                 }
             });
 
-            var descriptor = new RazorDiagnosticDescriptor(id, () => message, (RazorDiagnosticSeverity)severity);
-
+            var descriptor = CreateDiagnosticDescriptor(id, message, (RazorDiagnosticSeverity)severity);
             var diagnostic = RazorDiagnostic.Create(descriptor, sourceSpan);
             diagnostics.Add(diagnostic);
+
+            static RazorDiagnosticDescriptor CreateDiagnosticDescriptor(string id, string message, RazorDiagnosticSeverity severity)
+            {
+                // Do NOT inline this descriptor method into the above line.
+                //
+                // Reasoning:
+                //  Because "message" is referenced in not only the below "() => message" but also the above switch statement containing lambda without this separate static method the
+                //  compiler would generate a fake display class that looks something like:
+                //  private sealed class <>c__DisplayClass26_0
+                //  {
+                //      public JsonReader reader;
+                //      public string id;
+                //      public int severity;
+                //      public string message;
+                //      public SourceSpan sourceSpan;
+                //
+                //       internal void <ReadDiagnostic>b__0(string propertyName)
+                //       {
+                //           switch (propertyName)
+                //           {
+                //              ....
+                //               case "Message":
+                //                   if (reader.Read())
+                //                   {
+                //                       message = (string)reader.Value;
+                //                   }
+                //              ...
+                //           }
+                //       }
+                //
+                //       internal string <ReadDiagnostic>b__1()
+                //       {
+                //           return message;
+                //       }
+                //  }
+                //
+                // And then uses that display class's "b_1" method as the lambda parameter for the diagnostic descriptor. Its reasoning is to maintain correctness
```

Comment:
@jaredpar @333fred Is there not a better compiler optimization that can be done instead of re-using a display class? I understand the approach of the compiler to maintain correctness but dammmnn this was a shocker to find given its impact.

Reply:
Not really. We have no guarantees of when or where that other lambda will be run, and its reference to `message` has to update to the same content as this `message`.

Reply:
Would generating an in-the-middle DTO class or something that both reference not work?

Reply:
Potentially, but that's even more garbage, which is likely just as harmful in some scenarios as this is.

Reply:
Fair